packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs

```diff
@@ -0,0 +1,131 @@
+        // Warning: Only use `get_prop_unchecked` for getting a prop in this function.
+        // This is so that we possibly catch errors a little earlier if a prop that is not
+        // set in the dependency graph is accessed.
+        let node = data.node;
+        let document_model = data.document_model;
+        let origin = data.origin;
+
+        match self {
+            ContentFilter::IsType(type_name) => match node {
+                GraphNode::Component(_) => {
+                    let component_type = document_model.get_component_type(node);
+                    component_type.eq(type_name)
+                }
+                _ => false,
+            },
+            ContentFilter::HasPropMatchingProfile(profile) => match node {
+                GraphNode::Component(component_idx) => {
+                    let prop =
+                        document_model.get_component_prop_by_profile(component_idx, &[*profile]);
+                    prop.is_some()
+                }
+                _ => {
+                    // `PropProfile::String` is the only profile that non-component nodes can match.
+                    // We special case it here.
+                    if matches!(profile, PropProfile::String)
+                        && matches!(node, GraphNode::String(_))
+                    {
+                        return true;
+                    }
+                    false
+                }
+            },
+            ContentFilter::HasPropMatchingProfileAndCondition(profile, cond) => match node {
+                GraphNode::Component(component_idx) => {
+                    let prop =
+                        document_model.get_component_prop_by_profile(component_idx, &[*profile]);
+
+                    let prop = prop.map(|prop_pointer| {
+                        let prop_node = document_model.prop_pointer_to_prop_node(prop_pointer);
+                        document_model._get_prop_unchecked(prop_node, origin)
+                    });
+
+                    prop.map(|prop| {
+                        let value = prop.value;
+                        // Test the condition
+                        cond.apply_test(&value)
+                    })
+                    .unwrap_or(false)
+                }
+                _ => false,
+            },
+            ContentFilter::IsComponent => matches!(node, GraphNode::Component(_)),
+            ContentFilter::IsString => matches!(node, GraphNode::String(_)),
+        }
+    }
+
+    fn accumulate_deps(&self, data: &FilterData<'_>) -> Vec<GraphNode> {
+        // Warning: **Do not get prop values in this function**
+        // We won't initialize their deps here. Only return the required prop nodes.
+        let node = data.node;
+        let document_model = data.document_model;
+
+        match node {
+            GraphNode::Component(_) => {}
+            _ => {
+                // The only allowable deps are from props. If we are here, we're querying a non-component
```

Comment:
If we allow one to filter by the value of a string, as suggested above, then we'd need to add the string as a dependency, as we'd have to rerun the query if the value of the string changed.

Reply:
Is the mechanism to track string changes set up yet?

Reply:
Yep, it just works given that strings are stored in a `StringCache` that is based on the `PropCache`.

Reply:
The `<textInput>` tests involving actions modify string props and string children.